unity-client/Assets/Scripts/MainScripts/DCL/Controllers/HUD/EmailPromptHUD/EmailPromptHUDController.cs

```diff
@@ -1,34 +1,144 @@
 ﻿using UnityEngine;
-using UnityEngine.UI;
-using TMPro;
 using DCL.Interface;
+using DCL.Helpers;
+using System;
+using System.Collections;
+using DCL.Tutorial;
 
-public class EmailPromptHUDController : MonoBehaviour
+public class EmailPromptHUDController : IHUD
 {
-    public TMP_InputField inputField;
-    public GameObject cryptoUserMessage;
-    public GameObject nonCryptoUserMessage;
-    public Button sendButton;
+    const float POPUP_DELAY = 60;
 
-    void Awake()
+    EmailPromptHUDView view;
+
+    bool isRendererEnabled = false;
+    bool isTimerComplete = false;
+    bool isTimerCreated = false;
+    bool isEnable = false;
+
+    Coroutine timerRoutine;
+
+    public EmailPromptHUDController()
+    {
+        view = UnityEngine.Object.Instantiate(Resources.Load<GameObject>("EmailPromptHUD")).GetComponent<EmailPromptHUDView>();
+        view.name = "_EmailPromptHUD";
+
+        view.OnDismiss += OnDismiss;
+        view.OnSendEmail += OnSendEmail;
+
+        view.gameObject.SetActive(false);
+
+        CommonScriptableObjects.rendererState.OnChange += OnRendererStateChanged;
+    }
+
+    public void SetVisibility(bool visible)
+    {
+        if (visible)
+        {
+            Utils.UnlockCursor();
+            view.gameObject.SetActive(true);
+            view.showHideAnimator.Show();
+        }
+        else
+        {
+            view.showHideAnimator.Hide();
+        }
+    }
+
+    public void Dispose()
+    {
+        CommonScriptableObjects.rendererState.OnChange -= OnRendererStateChanged;
+
+        if (view != null)
+        {
+            GameObject.Destroy(view.gameObject);
+        }
+        if (timerRoutine != null)
+        {
+            CoroutineStarter.Stop(timerRoutine);
+        }
+    }
+
+    public void SetEnable(bool enable)
+    {
+        bool wasEnabled = isEnable;
+        isEnable = enable;
+
+        if (enable && !wasEnabled)
+        {
+            if (CommonScriptableObjects.rendererState)
+            {
+                isRendererEnabled = true;
+                CreateTimer();
+            }
+        }
+        else if (!enable && wasEnabled)
+        {
+            if (timerRoutine != null)
+            {
+                CoroutineStarter.Stop(timerRoutine);
+            }
+            isTimerComplete = false;
+            isTimerCreated = false;
+        }
+    }
+
+    void CreateTimer()
     {
-        bool hasWallet = UserProfile.GetOwnUserProfile().hasConnectedWeb3;
+        if (!isEnable)
+            return;
 
-        cryptoUserMessage.SetActive(hasWallet);
-        nonCryptoUserMessage.SetActive(!hasWallet);
+        isTimerCreated = true;
+        timerRoutine = CoroutineStarter.Start(WaitForSeconds(POPUP_DELAY));
```

Comment:
I have a few issues with the approach of this class:

* Too many flags: having logic based on a bunch of flags that can be true or false always lead to issues maintaining/reading the code down the road

* If we are already starting a coroutine here, and the code is dependant on it, why not just make the whole logic run on a single IEnumerator instead of all those flags flying around? like this:

```
IEnumerator StartEmailLogic_FindBetterName()
{
    yield return WaitForSecondsCache.Get(seconds);
    yield return WaitUntil( () => rendererIsEnabled );
    ShowEmailPopup();
}
```

Reply:
when you are right you are right 😄